### PR TITLE
3DS GPU rewrite

### PR DESCRIPTION
**Pull request by yuriks**
After much blood, sweat and tears. Replaces sf2d lib with custom code, which we can hopefully also scrap soon and replace with ctrulib-refactor and fincs-powered GPU code instead.

### CODE REVIEW

File: src/platform/3ds/CMakeLists.txt [hunk 1]
```diff
@@ -19,9 +19,23 @@ endif()
...
+	${CMAKE_CURRENT_BINARY_DIR}/font.c
+	${CMAKE_CURRENT_BINARY_DIR}/uishader.c
+	${CMAKE_CURRENT_BINARY_DIR}/uishader.h
+	${CMAKE_CURRENT_BINARY_DIR}/uishader.shbin.h
 
-set_source_files_properties(${CMAKE_CURRENT_BINARY_DIR}/font.c PROPERTIES GENERATED ON)
+	${CMAKE_CURRENT_SOURCE_DIR}/gui-font.c
+	${CMAKE_CURRENT_SOURCE_DIR}/ctr-gpu.c
+	${CMAKE_CURRENT_SOURCE_DIR}/ctr-gpu.h
+	)
```
Comment (endrift) [inline src/platform/3ds/CMakeLists.txt]: Paren shouldn't be on its own line (as per other code in this file).

File: src/platform/3ds/CMakeLists.txt [hunk 2]
```diff
@@ -38,6 +52,23 @@ add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/font.c
...
 
+add_custom_command(
+	OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/uishader.shbin ${CMAKE_CURRENT_BINARY_DIR}/uishader.shbin.h
+	MAIN_DEPENDENCY ${CMAKE_CURRENT_SOURCE_DIR}/uishader.vsh
+	COMMAND ${PICASSO}
+		-o ${CMAKE_CURRENT_BINARY_DIR}/uishader.shbin
+		-h ${CMAKE_CURRENT_BINARY_DIR}/uishader.shbin.h
+		${CMAKE_CURRENT_SOURCE_DIR}/uishader.vsh
+	COMMENT "picasso uishader.vsh"
+	)
```
Comment (endrift) [inline src/platform/3ds/CMakeLists.txt]: This should be a custom target. I can fix that after a merge if you don't want to deal with CMake though. Ditto about the paren though.

File: src/platform/3ds/CMakeLists.txt [hunk 3]
```diff
@@ -38,6 +52,23 @@ add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/font.c
...
+add_custom_command(
+	OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/uishader.c ${CMAKE_CURRENT_BINARY_DIR}/uishader.h
+	MAIN_DEPENDENCY ${CMAKE_CURRENT_BINARY_DIR}/uishader.shbin
+	COMMAND ${RAW2C} ${CMAKE_CURRENT_BINARY_DIR}/uishader.shbin
+	WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+	COMMENT "raw2c uishader.shbin"
+	)
```
Comment (endrift) [inline src/platform/3ds/CMakeLists.txt]: Ditto.

File: src/platform/3ds/gui-font.c [hunk 4]
```diff
@@ -3,35 +3,42 @@
  * This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
```
Comment (endrift) [inline src/platform/3ds/gui-font.c]: Any reason for the random newline here?

File: src/platform/3ds/gui-font.c [hunk 5]
```diff
@@ -3,35 +3,42 @@
...
-	sf2d_texture* tex;
+	struct ctrTexture texture;
 };
 
 struct GUIFont* GUIFontCreate(void) {
 	struct GUIFont* guiFont = malloc(sizeof(struct GUIFont));
-	if (!guiFont) {
-		return 0;
+	if (guiFont == NULL) {
+		return NULL;
```
Comment (endrift) [inline src/platform/3ds/gui-font.c]: No real reason to change this; it's equivalent to what it was, and I don't like code churn without functional changes.

File: src/platform/3ds/gui-font.c [hunk 6]
```diff
@@ -3,35 +3,42 @@
...
 	}
-	guiFont->tex = sf2d_create_texture(256, 128, TEXFMT_RGB5A1, SF2D_PLACE_RAM);
-	memcpy(guiFont->tex->data, font, font_size);
-	guiFont->tex->tiled = 1;
+
+	struct ctrTexture* tex = &guiFont->texture;
+	ctrTexture_Init(tex);
+	tex->data = linearAlloc(256 * 128 * 2);
```
Comment (endrift) [inline src/platform/3ds/gui-font.c]: Any reason this can't be in VRAM?

File: src/platform/3ds/main.c [hunk 7]
```diff
@@ -46,23 +46,29 @@ static struct GBAAVStream stream;
...
 static void _postAudioBuffer(struct GBAAVStream* stream, struct GBAAudio* audio);
 
 static void _drawStart(void) {
+	ctrGpuBeginFrame();
 	if (screenMode < SM_PA_TOP) {
-		sf2d_start_frame(GFX_BOTTOM, GFX_LEFT);
+		ctrSetViewportSize(320, 240);
 	} else {
-		sf2d_start_frame(GFX_TOP, GFX_LEFT);
+		ctrSetViewportSize(400, 240);
```
Comment (endrift) [inline src/platform/3ds/main.c]: Does ctrulib not have constants for the screen sizes?

File: src/platform/3ds/main.c [hunk 8]
```diff
@@ -117,64 +123,104 @@ static void _gameUnloaded(struct GBAGUIRunner* runner) {
...
 		break;
 	case SM_AF_TOP:
-		sf2d_draw_texture_scale_blend(tex, 20, 384, 1.5, -1.5, 0xFFFFFF3F | (faded ? 0 : 0xC0));
+		w = 360;
+		h = 240;
 		break;
 	case SM_AF_BOTTOM:
-		sf2d_draw_texture_scale_blend(tex, 0, 368 - 40 / 3, 4 / 3.0, -4 / 3.0, 0xFFFFFF3F | (faded ? 0 : 0xC0));
+		w = 318;
+		h = 212;
```
Comment (endrift) [inline src/platform/3ds/main.c]: This could use a comment to explain why these exact constants.

File: src/platform/3ds/main.c [hunk 9]
```diff
@@ -305,9 +351,19 @@ int main() {
...
+	if (ctrInitGpu() < 0) {
+		fprintf(stderr, "GPU init failed\n");
+	}
+
+	ctrTexture_Init(&gbaOutputTexture);
+	gbaOutputTexture.format = GPU_RGB565;
+	gbaOutputTexture.filter = GPU_LINEAR;
+	gbaOutputTexture.width = 256;
+	gbaOutputTexture.height = 256;
+	gbaOutputTexture.data = linearAlloc(256 * 256 * 2);
```
Comment (endrift) [inline src/platform/3ds/main.c]: Ditto about VRAM.

File: src/platform/3ds/main.c [hunk 10]
```diff
@@ -305,9 +351,19 @@ int main() {
 		audioRight = linearMemAlign(AUDIO_SAMPLE_BUFFER * sizeof(int16_t), 0x80);
 	}
 
-	sf2d_init();
-	sf2d_set_clear_color(0);
-	tex = sf2d_create_texture(256, 256, TEXFMT_RGB565, SF2D_PLACE_VRAM);
+	gfxInit(GSP_BGR8_OES, GSP_BGR8_OES, false);
+
+	if (ctrInitGpu() < 0) {
+		fprintf(stderr, "GPU init failed\n");
```
Comment (endrift) [inline src/platform/3ds/main.c]: Use `GBALog` instead.

File: src/platform/3ds/main.c (hunk 7 above)
Comment (shinyquagsire23) [inline src/platform/3ds/main.c]: @endrift It doesn't as far as I can tell.

Comment (yuriks) [inline src/platform/3ds/main.c]: Yeah, not as far as I can tell. (It does give you the LCD fb dimensions when you do gfxGetFramebuffer, but there's no reason to call that here.)

**Comment by yuriks** [comment]
Updated

**Comment by endrift** [comment]
Looks good. I'll test it on my N3DS when I get home, although some of these commits should be squashed before I merge.

**Pull request merged**